### ISSUE TITLE
Allow more C99 features

### ISSUE THREAD

**Issue by senioria** [vim/vim#11157]
It's said in the document that some C99 features are not used in Vim:

https://github.com/vim/vim/blob/b2f0ca820eae50994745106d824e215d87bd7926/runtime/doc/develop.txt#L218-L227

However, it may be possible to loose some of these restrictions:

- After dropping support for pre-Vista Windows systems, all Windows Vim users can use a newer version of MSVC (MSVC 2013) to compile code with declaration after statements. Besides, not allowing declarations after statements had caused some issues in Linux kernel; so it may be reasonable to also drop support to old MSVC, and allow this feature;
- VLAs are widely disallowed; maybe a reason other than compiler support can be used;
- [The thread reporting that flexible array members are not supported by HP-UX C compiler](https://groups.google.com/g/vim_dev/c/c6p9CXp6Zjw/m/dDs-QihHCQAJ) is 5 years old, and even at that time, it's a legacy problem; so this restriction may safely be removed now.

**Comment by brammool** [comment]
> It's said in the document that some C99 features are not used in Vim:
> 
> https://github.com/vim/vim/blob/b2f0ca820eae50994745106d824e215d87bd7926/runtime/doc/develop.txt#L218-L227
> 
> However, it may be possible to loose some of these restrictions:
> 
> - After dropping support for pre-Vista Windows systems, all Windows
> Vim users can use a newer version of MSVC (MSVC 2013) to compile code
> with declaration after statements. Besides, not allowing declarations
> after statements had caused some issues in Linux kernel; so it may be
> reasonable to also drop support to old MSVC, and allow this feature;

We can see if this will work now.  I do know that in the past one or two
of these new style declarations slipped through and some people
complained.  I'm not sure if that is related to dropping XP support,
probably not.  If a newer MSVC version is used then we'll find the next
compiler where it fails...

I actually like declaring all variables at the start of the block, that
way they are easy to find.  The main advantage of C99 is being able to
declare the loop variable inside "for".  We could use that in a couple
of places and see who has a problem with it.

> - VLAs are widely disallowed; maybe a reason other than compiler
> support can be used;

Variable length arrays by themselves are not all that useful.  We have
the garray_T which works well, because it stores the size as well.  And
allows for dynamic growing as well.

> - [The thread reporting that flexible array members are not supported
> by HP-UX C
> compiler](https://groups.google.com/g/vim_dev/c/c6p9CXp6Zjw/m/dDs-QihHCQAJ)
> is 5 years old, and even at that time, it's a legacy problem; so this
> restriction may safely be removed now.

Flexible array members are useful.  We actually use them but with old
style syntax (using an array of size one that is actually bigger).

That thread is only four years old, and this type of machine running
HP/UX tends to keep running for a very long time.  Since we already have
an implementation that works, and there is hardly an advantage other
than "let's use this cool feature", I don't think there is much point in
changing.

-- 
From "know your smileys":
 :.-(	Crying

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by 91khr** [comment]
> We can see if this will work now.  I do know that in the past one or two
of these new style declarations slipped through and some people
complained.  I'm not sure if that is related to dropping XP support,
probably not.  If a newer MSVC version is used then we'll find the next
compiler where it fails...

It's mentioned in the comment of the newly-approved pr dropping XP support that enabling users to compile Vim on their
own may also be a concern, and the first version of MSVC that supports the feature is MSVC 2013, which requires Windows
7. So if users being able to compile is important, only after dropping pre-Vista support can this restriction be
removed.

And after #11183, which drops support for all pre-2015 MSVC, there may be no concern removing the restriction.

> The main advantage of C99 is being able to
declare the loop variable inside "for".  We could use that in a couple
of places and see who has a problem with it.

Seems that this is the reason why Linux upgraded their language to C11; but they uses macros to generate for loops,
while it seems that loops in Vim are mostly just C's builtin fors and whiles.
So in Vim the problem may not be so urgent.

> Since we already have
an implementation that works, and there is hardly an advantage other
than "let's use this cool feature", I don't think there is much point in
changing.

Yes this is true...

**Comment by brammool** [comment]
The newly used features have not resulted in complaints, thus it appears this work for all places where Vim is build.

I don't think there is a good reason to add more C99 features and risking build breakage right now.
If there is a specific feature that would make the code so much cleaner it's worth taking a risk, plese
open a separate issue for each.  Closing this one now.

**Comment by dvogel** [comment]
Catching this a little late, but I wanted to mention I did some research on this front a few months ago. I'll share a few documentation resources that may be helpful references as this shakes out.

VMS Software confirmed that they have full C99 support. From [twitter](https://twitter.com/VMSSoftware/status/1545424509372596227):

> John Reagan, our Master Compiler Engineer, replies, "The compiler has been C99 compliant for years. However, the run-time environment and headers have only recently been made C99 compliant. You need RTL ECO05 on Alpha and Itanium. V9.2 x86 will be C99 compliant at first ship."

So if anyone using VMS complains it's likely they aren't running the most current VMS version.

I was unable to confirm first hand or through HP support but even HP-UX appears to have [C99 support for flexible array members](https://support.hpe.com/hpesc/public/docDisplay?docId=pdbPHSS_41187&docLocale=en_US) though users where may need to use `c99` instead of `cc`.

Finally, one useful MSVC-related resource I had found was only available [via the Wayback Machine](https://web.archive.org/web/20130701001452/http://blogs.msdn.com/b/vcblog/archive/2013/06/28/c-11-14-stl-features-fixes-and-breaking-changes-in-vs-2013.aspx).
